cpp/src/arrow/compute/kernels/scalar_nested_test.cc

```diff
@@ -116,6 +117,111 @@ TEST(TestScalarNested, ListElementInvalid) {
               Raises(StatusCode::Invalid));
 }
 
+TEST(TestScalarNested, ListSliceVariableOutput) {
+  const auto value_types = {float32(), int32()};
+  for (auto value_type : value_types) {
+    /* Variable list size output required variable size list input. */
+    auto inputs = {ArrayFromJSON(list(value_type), "[[1, 2, 3], [4, 5], [6], null]")};
+    for (auto input : inputs) {
+      ListSliceOptions args(/*start=*/0, /*stop=*/2, /*step=*/1,
+                            /*return_fixed_size_list=*/false);
+      auto expected = ArrayFromJSON(list(value_type), "[[1, 2], [4, 5], [6], null]");
+      CheckScalarUnary("list_slice", input, expected, &args);
+
+      args.start = 1;
+      expected = ArrayFromJSON(list(value_type), "[[2], [5], [], null]");
+      CheckScalarUnary("list_slice", input, expected, &args);
+
+      args.start = 2;
+      args.stop = 4;
+      expected = ArrayFromJSON(list(value_type), "[[3], [], [], null]");
+      CheckScalarUnary("list_slice", input, expected, &args);
+    }
+  }
+
+  // Verify passing `return_fixed_size_list=false` with fixed size input
+  // returns variable size even if stop is beyond list_size
+  ListSliceOptions args(/*start=*/0, /*stop=*/2, /*step=*/1,
+                        /*return_fixed_size_list=*/false);
+  auto input = ArrayFromJSON(fixed_size_list(int32(), 1), "[[1]]");
+  auto expected = ArrayFromJSON(list(int32()), "[[1]]");
```

Comment:
Do we need to support this? I find it a bit strange that someone would ever want to slice an already fixed size list array (so slicing will always give list elements of all the same size) as a non-fixed size list.

Reply:
I agree, _probably_ not used or at least that often. But if it's "just works" by the implementation, is there a reason we should actively prevent it? Can certainly throw in a check for this scenario and raise an error.